impute_constant/impute_constant.tex

```diff
@@ -0,0 +1,196 @@
+    
+    def Relation(d_in: u32, d_out: u32) -> bool: |\label{line:rel}|
+        return d_out >= d_in*1
+
+    def function(data: Vec[DA::Carrier]) -> Vec[DA::NonNull]: |\label{line:fn}|
+        def impute_constant(x: DA) -> DA::NonNull: |\label{line:atom-fn}|
+            return constant if x.is_null else x
+        return list(map(impute_constant, data)) |\label{line:map}|
+        
+    # can we comment out input_metric, output_metric, and stability_relation, and function if it's not being called anymore?
+    return make_row_by_row(input_domain, output_domain, impute_constant); |\label{line:row-transform}|
+\end{lstlisting}
+
+\subsection{Proof}
+
+\begin{lemma}[\texttt{DA::NonNull} contains \texttt{null}] \label{lemma:NonNull}
+A \texttt{var} of type \texttt{DA::NonNull} can be of type \texttt{null}.
+\end{lemma}
+\begin{proof}
+Let the domain of atom variable \texttt{DA} be \texttt{InherentNullDomain<AllDomain<f64>>}. Recall that \texttt{InherentNullDomain} exists for types that can represent null inherently in the carrier type. Then the type $$\texttt{DA::NonNull == InherentNullDomain<AllDomain<f64>>::NonNull == f64}.$$ The latter holds because in the \texttt{InherentNullDomain} implementation in the rust code \url{https://github.com/opendp/opendp/blob/main/rust/opendp/src/trans/impute.rs#L48-L56}, the \texttt{type NonNull = Self::Carrier}. The \texttt{Carrier} of \texttt{VectorDomain<AllDomain<T>>} has type \texttt{T}, so in this case the \texttt{::Carrier} is type \texttt{f64}.
+
+Therefore \texttt{var} is also of type \texttt{f64}. \texttt{f64} can contain \texttt{null} values, so we are done.
+\end{proof}
+
+\begin{lemma}[Precondition for row transform]\label{lemma:pure-fn}
+\texttt{make\_impute\_constant} satisfies the preconditions of \texttt{make\_row\_by\_row}. That is, the atom function \texttt{impute\_constant} in pseudo code line \ref{line:atom-fn} is a pure function. 
+\end{lemma}
+\begin{proof}
+We reference the examples in \href{https://en.wikipedia.org/wiki/Pure_function}{Pure function Wikipedia page} for the definition of a pure function, and later as a check list to verify the properties of a pure function hold. To verify whether \texttt{impute\_constant} is a pure function, it must satisfy the following properties:
+\begin{enumerate}
+    \item The function return values are identical for identical arguments.
+    \item The function application has no side effects.
+\end{enumerate}
+
+The first property is satisfied because there is no static variable defined within \\\texttt{make\_impute\_constant}, no mutable reference argument, no return value of an input stream, all of which could potentially cause different function outputs for inputs. Even though the function returns a non-local variable \texttt{constant}, \texttt{constant} is never changed within the code of \texttt{make\_impute\_function}, so the return value stays the same for same input to \texttt{impute\_constant}, all of which could produce side effects.
```

Comment:
please ask Marco Gaboardi to review this when he comes back

Reply:
I don't think "constant" is a variable anymore once we've defined the function make_impute_constant; it is now hardwired as a constant in the function

Reply:
do you mean "none of which could produce side effects"?